tempesta_fw/classifier/frang.c

```diff
@@ -502,14 +537,38 @@ frang_http_host_check(const TfwHttpReq *req, FrangAcc *ra)
 	return ret;
 }
 
+static int
+frang_resp_time_in_frame(const int ta, const int tb)
+{
+	return tb - ta + (ta > tb ? FRANG_RESP_TIME_PERIOD : 0)
+		< frang_cfg.http_resp_code_block->tf;
+}
+
+static int
+frang_bad_resp_limit(FrangAcc *ra)
+{
+	FrangRespCodeStat *stat = ra->resp_code_stat;
+	long cnt = 0;
+	const int ts = (jiffies / HZ) % FRANG_RESP_TIME_PERIOD;
```

Comment:
You have added a few restrictions which only complicates code. Try to keep them as simpler as possible. Having `ts` of the same type as `jiffiies` wouldn't require `FRANG_RESP_TIME_PERIOD` macro. You don't need to check if the requests are made in the same year,  keep track of integer overflow and all that stuff.

I've looked though the latest changes (FrangRates history approach) and i find usage of `FrangRespCodeStat.ts` (and `i` from line below) to be wrong. Let's look to the idea behind the `FrangRespCodeStat.ts` member and the `FrangAcc.FrangRespCodeStat[FRANG_FREQ])` circular buffer. Instead of working with continuous time we measure time with small discrete quantum. Size of each quantum is `(timeframe in seconds) /   FRANG_FREQ`. The `FrangRespCodeStat.ts` is time measured in such quantum.

On processing response we get current system time in jiffies and translate to that quantum. Then we get the number of the quantum in current time frame. This tells which exactly array member of  `FrangAcc.FrangRespCodeStat[FRANG_FREQ])` must be used. If timestamp of the current `FrangRespCodeStat` doesn't equals to current timestamp, we can simply overwrite it since it out of current time frame. Finally we increment number of responses for the current `FrangRespCodeStat`.

On processing request we also get current system time in jiffies and translate to that quantum. That is needed to find beginning of the current time frame. Then we traverse over the `FrangAcc.FrangRespCodeStat[FRANG_FREQ])` circular buffer to get number of responses in the current time frame (time stamp is bigger or equal to the beginning of current time frame).

What is needed to be done:
- `FrangRespCodeStat.ts` must be type of `unsigned long` (same as `jiffies`)
- introduce function to convert jiffies to current quantum depending of time frame size. The conversion is used 3 times: `frang_bad_resp_limit()`, `frang_resp_handler()` and `frang_req_limit()` (the latest uses 1 second time frames). 
`ts = jiffies * FRANG_FREQ / (HZ * time_frame_size)`
- current member in circular buffer is directly connected to the time stamp: 
`i = ts % (FRANG_FREQ * time_frame_size)`
- introduce helper function (or macro) for collecting current rate (twice:  `frang_bad_resp_limit()`, and `frang_req_limit()`) to compare timestamp of buffer member to start of current timestamp:
`if (buff[i].ts + FRANG_FREQ * time_frame_size >= ts) ...`